Speed up editoast errors db insertion

flomonster opened the pull request:
## What this PR does

Editing an object takes too much time because we have to refresh the whole error layer each time.
The idea is to identify the new errors and those that have disappeared, once all the new errors have been generated. This avoids deleting everything and inserting all the new errors because in practice there are few changes.
To achieve this, we hash (with sha1) the errors.  

## Benchmark

|                | Ref    | New  |
|----------------|--------|------|
| Move detector  | 4089ms | 11ms |
| Remove a track | 3881ms | 89ms |
| Auto fix       | 4225ms | 30ms |

This benchmark was done using a big infra (France) with ~10,000 errors.
In practice editing a big infra took at least `5s` now it takes `1.5s`.

close #5516

flomonster committed:
editoast: speed up errors db insertion

flomonster commented:
## Unicity error

I haven't checked why yet but the unicity constraint on the `info_hash` can fail. Two possibilities:
- Our error generator can generate duplicates
- My PR is broken and I'm trying to add an existing error

younesschrifi reviewed (commented):
LGTM
I'm investigating the unicity error.

flomonster committed:
editoast: fix linear object error level

flomonster commented:
I fixed the unicity error. It was not an error in error generators but in the error layer generator.
Operational point errors were duplicated since their geometry is split among multiple rows in DB (needed to have multiple kp). I fixed it using `ST_Collect` so now one error generates only one row in DB.

younesschrifi reviewed (approved):
very nice PR !! LGTM ! 
I tested it and the speed has increased considerably.

woshilapin commented on editoast/migrations/2023-11-24-162712_add_errors_hash/up.sql:
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE infra_layer_error ADD info_hash varchar(40) NOT NULL DEFAULT 'undefined';
+UPDATE infra_layer_error SET info_hash = id::text;
```
Isn't it possible to initialize the column with a SHA1 already ([the `digest` function](https://www.postgresql.org/docs/current/pgcrypto.html#PGCRYPTO-GENERAL-HASHING-FUNCS-DIGEST) can do that)? Note that at the moment, you serialize the `struct` as JSON which might not be possible to easily reproduce in SQL, but it doesn't make too much sense to use JSON to hash.

flomonster replied:
I didn't want to do this for multiple reasons:
- It will be very difficult to reproduce the same hash that editoast will produce (it depends on how the JSON is serialized)
- The first edition will be slower cause all errors will be removed and inserted with the correct hash. But it's no worse than what we already have. 
- We can think of it as a poorly initialized cache

For me, it was an acceptable compromise to avoid bothering with migration.

woshilapin commented on editoast/src/generated_data/error/mod.rs:
```diff
@@ -87,6 +92,23 @@ impl<Ctx> GlobalErrorGenerator<Ctx> {
 }
 
+/// An infra error with its hash
+struct ErrorWithHash {
+    error: InfraError,
+    /// Sha1 of the serialized error
+    hash: String,
+}
+
```
Instead of doing a wrapper type of `InfraError` with the additional field, why not just implementing `Hash` directly on `InfraError`? It would avoid a new type and use the standard traits for that kind of usecase.

flomonster replied:
We can't use the `std::Hash` trait with sha1. 
I don't want to use the `std::collections::hash_map::DefaultHasher` since there is no guarantee of the algorithm used and therefore of the hash produced.

------

I can still implement a function to retrieve the sha1 of an error. Therefore I wanted to compute it only once since it requires serializing `information`. This is why I created this wrapper.

woshilapin commented on editoast/src/generated_data/error/mod.rs:
```diff
@@ -246,30 +268,117 @@ fn get_insert_errors_query(obj_type: ObjectType) -> &'static str {
+async fn retrieve_current_errors_hash(
+    conn: &mut PgConnection,
+    infra_id: i64,
+) -> Result<Vec<String>> {
```
Would it make sense to return a more specific type than `String`? Like a specific `Hash` type, or at least a `NonBlankString`?

flomonster replied:
We could use `[u8; 20]` instead of `String` but this means we expect a valid sha1 value in DB.
I can define a `type ErrorHash = String` for now so it will be easier to change it later.

woshilapin commented on editoast/src/generated_data/error/mod.rs:
```diff
@@ -246,30 +268,117 @@ fn get_insert_errors_query(obj_type: ObjectType) -> &'static str {
     for error in errors {
-        errors_by_type
-            .entry(error.get_type())
+        errors_by_layer
```
I'm not sure that I understand the change of naming for this variable since the key in the `HashMap` is `error.get_type()`, therefore, `errors_by_type` seems more intuitive. And you even seems to use it as `errors_by_type` in some places (see [here](https://github.com/osrd-project/osrd/pull/5868/files#diff-a5d39d0f5d978a1f0ba1662039ef9762984a0f3aea45462d8d5ff4a89a6e1514R322)). But I'm probably missing some context here.

woshilapin commented on editoast/src/generated_data/error/mod.rs:
```diff
@@ -246,30 +268,117 @@ fn get_insert_errors_query(obj_type: ObjectType) -> &'static str {
+    )
+    .execute(conn)
+    .await?;
+    assert_eq!(nb_deleted, errors_hash.len());
```
Is it ok to panic in production? You may prefer `debug_assert_eq!()` instead (see [documentation](https://doc.rust-lang.org/std/macro.debug_assert_eq.html))?

woshilapin reviewed (changes requested):
Really nice performance improvements. No major comment, mostly naming questions and challenging some of the design decisions.

flomonster committed:
fixup! editoast: speed up errors db insertion

Tguisnet reviewed (approved):
Very cool PR, thanks.
LGTM (and tested)
